packages/plugin-pre/src/PreDomRenderer.ts

```diff
@@ -0,0 +1,53 @@
+import { AbstractRenderer } from '../../plugin-renderer/src/AbstractRenderer';
+import { DomRenderingEngine } from '../../plugin-dom/src/DomRenderingEngine';
+import { PreNode } from './PreNode';
+
+export class PreDomRenderer extends AbstractRenderer<Node[]> {
+    static id = 'dom';
+    engine: DomRenderingEngine;
+    predicate = PreNode;
+
+    /**
+     * Render the VNode to the given format.
+     */
+    async render(node: PreNode): Promise<Node[]> {
+        const pre = document.createElement('pre');
+        this.engine.renderAttributes(node.attributes, pre);
+        for (const child of node.children()) {
+            let domChildren = await this.renderChild(child);
+            // TODO: prevent CharNode from rendering spaces as nbsp.
+            pre.append(...this._brToNewline(domChildren));
+        }
+        return [pre];
+    }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Replace every <br> in a list of DOM Nodes with newlines.
+     *
+     * @param domNodes
+     */
+    _brToNewline(domNodes: Node[]): Node[] {
+        if (domNodes.length > 1) {
+            // Remove the last <br> if it's a placeholder (a trailing BR after
+            // another <br>, added just to make the first one visible).
```

Comment:
This is bad design on our part (somewhat) because we're manually undoing the work that we manually did in the `LineBreakDomRenderer`. I guess we should have a `PreSeparatorDomRenderer` which predicate is:
```node.ancestor(PreNode) && node instanceof this.engine.editor.configuration.defaults.Separator```
and never let `LineBreakDomRenderer` intervene here.

Reply:
If it's parsed as the character `\n` then it's not a br and does not need to be converted now, does it ?

Reply:
Right :duck:

Reply:
https://github.com/odoo-dev/jabberwock/pull/162/commits/35c8ac3640c40b884d8189a8b1ee9026a060ada5